### PR TITLE
NO_{GLYPH->DATA}_RENDERER warning

### PR DESCRIPTION
**Pull request by brendancol**
This PR addresses two current issues:
https://github.com/bokeh/bokeh/issues/3223
https://github.com/bokeh/bokeh/issues/3302

In the past, any valid plot required a `GlyphRenderer`.  That is no longer case with the addition of `TileRenderer` and `DynamicImageRenderer`.  The PR adds a new abstract base class `DataRenderer` to `bokeh/models/renderers.py` which inherits from `Renderer` and is now the parent class of `GlyphRenderer`, `TileRenderer`, and `DynamicImageRenderer`.  This parent class can now be validated in `bokeh/models/plots.py`. Initial testing looks fine and this removes the warning displayed by the `dynamic_map.py` example.

There are also a couple flake8 fixes...

### CODE REVIEW

File: bokeh/models/renderers.py [hunk 1]
```diff
@@ -26,7 +26,14 @@ class Renderer(Model):
     generally useful to instantiate on its own.
 
     """
-class TileRenderer(Renderer):
+
+@abstract
+class DataRenderer(Renderer):
+    """ A base class for data renderer types (e.g. GlyphRenderer). ``DataRenderer`` is not
+    generally useful to instantiate on its own.
```
Comment (mattpap) [inline bokeh/models/renderers.py]: No. Abstract classes can not be instantiated, but I see where this comes from. Please change `Renderer`'s docstring as well. "An abstract base class for ..." should suffice. Also, `GlyphRenderer` should be enclosed with double backticks.

**Comment by mattpap** [comment]
Looks good. Tests failed due to phantomjs crash.

Great you improved the quality of the source code involved, however you should have approached this differently for three reasons. First, it's harder than necessary for a reviewer to separate meaningful changes from refactoring. Secondly, if the PR stalls or there are lots of merges in the meantime, you will have a lot of really boring work resolving conflicts. Lastly, you are at risk of stalling a PR because we will be arguing about stylistic issues. My suggestion, at least separate refactoring from important changes in different commits. Preferably, any large scale refactoring should go through a separate PR. In this particular case, first option would suffice.

**Comment by brendancol** [comment]
@mattpap great explanation. moving flake8 fixes to separate PR.

**Comment by mattpap** [comment]
@brendancol, that wasn't necessary in this PR, I was just talking about future PRs. But OK, lets proceed.

**Comment by brendancol** [comment]
Thanks to BeyondCompare it was no problem.

**Comment by bryevdv** [comment]
LGTM, tested locally, will merge when travis passes and report inspected.

**Pull request merged**